mcstatus/protocol/connection.py

```diff
@@ -221,119 +525,141 @@ def read(self, length: int) -> bytearray:
+        """Read up to length bytes from self.reader."""
         result = bytearray()
         while len(result) < length:
             new = await asyncio.wait_for(self.reader.read(length - len(result)), timeout=self.timeout)
             if len(new) == 0:
-                raise IOError("Server did not respond with any information!")
+                raise IOError("Socket did not respond with any information!")
             result.extend(new)
         return result
 
-    def write(self, data: Union[bytes, bytearray]) -> None:
+    def write(self, data: Union[Connection, str, bytes, bytearray]) -> None:
+        """Write data to self.writer."""
+        if isinstance(data, Connection):
+            data = bytearray(data.flush())
+        elif isinstance(data, str):
+            data = bytearray(data, "utf-8")
         self.writer.write(data)
 
-    def __del__(self):
-        try:
-            self.writer.close()
-        except Exception:  # TODO: Check what this actually expects
-            pass
+    def close(self) -> None:
+        """Close self.writer."""
+        if self.writer is not None:  # If initialized
+            try:
+                self.writer.close()
+            except RuntimeError:  # There is a case where event loop is closed
+                pass
 
+    def __del__(self) -> None:
+        """Close self."""
+        self.close()
 
-class UDPAsyncSocketConnection(AsyncReadConnection):
-    # These will only be None until connect is called, ignore the None type assignment
-    stream: asyncio_dgram.aio.DatagramClient = None  # type: ignore[assignment]
-    timeout: float = None  # type: ignore[assignment]
 
-    def __init__(self):
-        super().__init__()
+class UDPAsyncSocketConnection(BaseAsyncConnection):
+    """Asynchronous UDP Connection class"""
+
+    __slots__ = ("stream", "timeout")
+
+    def __init__(self) -> None:
+        # This will only be None until connect is called, ignore the None type assignment
+        self.stream: asyncio_dgram.aio.DatagramClient = None  # type: ignore[assignment]
+        self.timeout: float = 0.0
 
     async def connect(self, addr: Address, timeout: float = 3) -> None:
+        """Connect to address. Timeout is in seconds."""
         self.timeout = timeout
         conn = asyncio_dgram.connect((addr[0], addr[1]))
         self.stream = await asyncio.wait_for(conn, timeout=self.timeout)
 
-    def flush(self) -> bytearray:
-        raise NotImplementedError("UDPSocketConnection does not support flush()")
-
-    def receive(self, data: Union[SupportsBytes, BytesConvertable]) -> None:
-        raise NotImplementedError("UDPSocketConnection does not support receive()")
-
     def remaining(self) -> int:
+        """Return 65535 (2 ** 16 - 1)."""
         return 65535
 
-    async def read(self, length: int) -> bytes:
+    async def read(self, length: int) -> bytearray:
+        """Read from stream. Length does nothing."""
+        # pylint: disable=unused-variable
```

Comment:
Are we using pylint?

Reply:
No, but I made an IDLE extension for pylint that I've been using while writing this.

Reply:
Removed as of a09673e